### PR TITLE
Generate QR codes for passwords

### PR DESCRIPTION
**Pull request by kaHaleMaKai**
Hi carnager,

I have added a very simple way of generating QR codes for passwords. I find it very handy when _transferring_ passwords to my phone. Please check if my PR meets your standards and needs, and provide feedback otherwise.

Best and thanks again for your awesome piece of code,
Lars

### CODE REVIEW

**Comment by carnager** [comment]
I like the idea. But the program to display the qrcode should be customizable, imho

**Comment by kaHaleMaKai** [comment]
We could with something like this (for image viewers reading from stdin):
```bash
qrencoder='qrencode -d 300 -v 8 -l H -o -'
image_viewer='feh -'

pass "$selected_password" | head -n 1 | $qrencoder | $image_viewer
```
Thus no intermediate saving of the qrencoded passwort. What do you think?

File: rofi-pass [hunk 1]
```diff
@@ -9,6 +9,12 @@ _rofi () {
     rofi -no-auto-select -i "$@"
 }
 
+has_program() {
```
Comment (moviuro) [inline rofi-pass]: This should totally use exit codes:

```
has_program() {
  command -v "$1" >/dev/null 2>&1
}
```

File: rofi-pass [hunk 2]
```diff
@@ -26,6 +32,8 @@ clip=primary
 default_user="${ROFI_PASS_DEFAULT_USER-$(whoami)}"
 default_user2=john_doe
 password_length=12
+has_qrencode="$(has_program qrencode)"
```
Comment (moviuro) [inline rofi-pass]: This shouldn't exist, as we can do without

File: rofi-pass [hunk 3]
```diff
@@ -308,6 +323,11 @@ mainMenu () {
     -kb-custom-16 "${help}"
     -kb-custom-17 "${switch}"
     -kb-custom-18 "${insert_pass}"
+    )
+    if [[ "$has_qrencode" -eq 1 ]] && [[ "$has_feh" -eq 1 ]]; then
```
Comment (moviuro) [inline rofi-pass]: Becomes `if has_program feh && has_program qrencode; then`

**Comment by carnager** [comment]
I came up with this...

```# internal whitelist of supported viewers
image_viewers_whitelist=("feh" "sxiv" "display")

# function to check user setting for viewer against whitelist
image_viewer_match () {
  local e match="$1"
  shift
  for e; do [[ "$e" == "$match" ]] && return 0; done
  return 1
}

set_image_viewer() {
  ## check if user defined viewer is supported
  if [[ -z $img_viewer ]]
  then
    image_viewer_match "${img_viewer}" "${image_viewers_whitelist[@]}"
    if [[ $? -eq "0" ]]
    then
      image_viewer="${img_viewer}"
    else
      ## check available supported viewers and use first match
      for i in "${image_viewers_whitelist[@]}"
      do
        command -v "$i" >/dev/null 2>&1
        if [[ $? -eq 0 ]]
        then
          image_viewer="${i}"
          break                                                                                                                     
        fi
      done
    fi
  fi
  if [[ -z $image_viewer ]]
  then
      echo "No supported image viewer found"
  fi
}
```
[...]
```
generateQrCode() {
  set_image_viewer                                                                                                                   
  if [[ -z $image_viewer ]]
  then
    printf '%s\n' "${image_viewers_whitelist[@]}" | _rofi -mesg "No supported viewer found. Supported viewers listed below" -dmenu
    exit_code=$?
    if [[ $exit_code -eq "1" ]]
    then
        exit
    fi
    "${basecommand}"
  fi
  checkIfPass
  pass "$selected_password" |head -n 1| qrencode -d 300 -v 8 -l H -o -|$image_viewer -
  clearUp
}
```
generateQrCode() {
  set_image_viewer
  if [[ -z $image_viewer ]]
  then
    printf '%s\n' "${image_viewers_whitelist[@]}" | _rofi -mesg "No supported viewer found. Supported viewers listed below" -dmenu
    exit_code=$?
    if [[ $exit_code -eq "1" ]]
    then
        exit
    fi
    "${basecommand}"
  fi
  has_qrencode
  if [[ $? -eq "1" ]]
  then
    echo "qrencode not found" | _rofi -dmenu
    exit_code=$?
    if [[ $exit_code -eq "1" ]]
    then
      exit
    fi
  fi
  checkIfPass
  pass "$selected_password" |head -n 1| qrencode -d 300 -v 8 -l H -o -|$image_viewer -                                              
  clearUp
}
```

**Comment by moviuro** [comment]
```
pass "$selected_password" |head -n 1| qrencode -d 300 -v 8 -l H -o -|image_viewer
```

`$image_viewer` should probably the function "image viewer that can display what it gets from a pipe":

```
if <image_viewer function is not yet defined>; then
  if has_command feh; then
    image_viewer() {
      #specific feh(1) syntax
      feh -
    }
  elif has_command sxiv; then
    image_viewer() {
      #specific sxiv syntax
      sxiv --file=- --no-full-screen # totally out of my hat, probably broken
    }
  ...
  else
    image_viewer() {
      # We didn't find a valid image viewer
      <print an error with rofi>
    }
  fi
fi
```

**Comment by carnager** [comment]
the rofi error message should only be displayed when using Alt+q, otherwise it's plain annoying.

**Comment by moviuro** [comment]
> only be displayed when using Alt+q

Defining a function, like I propose, does not run it. It will only be run if the user presses the key combo.

> which I really dislike

Then we don't provide a solution for this, and add a comment like *\# you really shouldn't use temporary files if you don't know what you're doing*.

**Comment by carnager** [comment]
This is a barebones solution for this mess. A simple function which the user can define however he likes. Same as is already done with the _rofi function. A comment that mentions the need for pipe support is also added. 

@moviuro @kaHaleMaKai what do you think?

https://github.com/carnager/rofi-pass/commit/04e61acfe6ae6792b96f640c3f7d4d2006cf9825

**Comment by carnager** [comment]
@kaHaleMaKai I suggest you update your original PR to reflect my qrcode branch, if you care about Contributor status.

**Comment by kaHaleMaKai** [comment]
@carnager sorry, was on sick leave. I don't care about the status plus apart from the original idea, it's not my code. All in all a fine piece of code, I think!

**Pull request closed without merging**